File: leetcode/4sum.py
# Given an array S of n integers, are there elements a, b, c, and d in S 
# such that a + b + c + d = target? Find all unique quadruplets in the array 
# which gives the sum of target.

class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[][]}
    def fourSum(self, nums, target):
        all_pairs = {}
        nums.sort()
        for i, j in ((s, l) for l in xrange(len(nums)) for s in xrange(l)):
            all_pairs.setdefault(nums[i] + nums[j], []).append((i, j))
        return [list(solution) for solution in
          set(((nums[i], nums[j], nums[k], nums[l])
          for j in xrange(len(nums)) for i in xrange(j)
          for k, l in all_pairs.get(target - nums[i] - nums[j], []) if k > j))]
                    
if __name__ == "__main__":
    solution = Solution()
    print solution.fourSum([1, 0, -1, 0, -2, 2], 0)

File: leetcode/4sum.js
/* Given an array S of n integers, are there elements a, b, c, and d in S 
 * such that a + b + c + d = target? Find all unique quadruplets in the array 
 * which gives the sum of target.
 */

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    unique = {}
    nums.sort(function(a,b){return a - b})
    for (var i = 0; i < nums.length - 3; ++i) {
        for (var j = i + 1; j < nums.length - 2; ++j) {
            var k = nums[i] + nums[j]
            for (var left = j + 1, right = nums.length - 1; left < right; ) {
                s = k + nums[left] + nums[right]
                if (s == target) {
                    unique[[nums[i], nums[j], nums[left], nums[right]]] = true
                    ++left
                    --right
                } else if (s < target)
                    ++left
                else
                    --right
            }
        }
    }
    result = []
    for (var key in unique) {
        var array = key.split(",")
        for (var i = 0; i < 4; ++i)
            array[i] = parseInt(array[i])
        result.push(array)
    }
    return result
};
